### PR TITLE
pubsub: panic if Ack is called twice on one message

### LINKED ISSUE #688: pubsub: specify behavior of calling Ack multiple times on a message
In the design doc and in the current set of stubs, `*Message.Ack`'s doc comment currently does not specify whether it is acceptable to call `Ack` multiple times on a single message. I would advise for it to be idempotent, much like [`context.CancelFunc`](https://godoc.org/context#CancelFunc).

### PR DESCRIPTION
**Pull request by ijt**
Fixes #688

### CODE REVIEW

File: internal/docs/pubsub/design.md [hunk 1]
```diff
@@ -371,13 +371,16 @@ type Message struct {
...
+	// isAcked tells whether the message has already had its Ack method called.
+	isAcked bool
 }
 
 type AckID interface{}
 
 // Ack acknowledges the message, telling the server that it does not need to
 // be sent again to the associated Subscription. This method returns
-// immediately.
+// immediately. Calling this method twice on the same message causes a panic.
```
Comment (zombiezen) [inline internal/docs/pubsub/design.md]:
```suggestion
// immediately. If Ack has already been called on the message, Ack panics.
```

File: internal/docs/pubsub/design.md [hunk 2]
```diff
@@ -371,13 +371,16 @@ type Message struct {
 
 	// sub is the Subscription this message was received from.
 	sub *Subscription
+
+	// isAcked tells whether the message has already had its Ack method called.
```
Comment (zombiezen) [inline internal/docs/pubsub/design.md]:
```suggestion
	// isAcked is true if Ack has been called on this message.
```

File: internal/pubsub/pubsub.go [hunk 3]
```diff
@@ -35,13 +36,21 @@ type Message struct {
...
+	// isAcked tells whether this message has already had its Ack method
+	// called.
+	isAcked bool
 }
 
 // Ack acknowledges the message, telling the server that it does not need to be
 // sent again to the associated Subscription. It returns immediately, but the
 // actual ack is sent in the background, and is not guaranteed to succeed.
 func (m *Message) Ack() {
+	if m.isAcked {
```
[the remark this answers is not in the document]
Comment (zombiezen) [reply internal/pubsub/pubsub.go]: `sync.Once` could also work.

[the remark this answers is not in the document]
Comment (ijt) [reply internal/pubsub/pubsub.go]: Done.

**Pull request merged**